tests/test_overdrive.py

```diff
@@ -1093,6 +1094,98 @@ def test_sync_bookshelf_ignores_holds_from_other_collections(self):
+        now = datetime.utcnow()
+
+        # The first time this Monitor is called, its 'start time' is
+        # the current time, and we ask for an overlap of one minute.
+        # This isn't very effective, but we have to start somewhere.
+        #
+        # (This isn't how the Overdrive collection is initially
+        # populated, BTW -- that's FullOverdriveCollectionMonitor.)
+        self.time_eq(start, now-monitor.OVERLAP)
+        self.time_eq(cutoff, now)
+        timestamp = monitor.timestamp()
+        eq_(start, timestamp.start)
+        eq_(cutoff, timestamp.finish)
+
+        # The second time the Monitor is called, its 'start time'
+        # is one minute before the previous cutoff time.
+        monitor.run()
+        new_start, new_cutoff, new_progress = monitor.catch_up_from_called_with
+        now = datetime.utcnow()
+        eq_(new_start, cutoff-monitor.OVERLAP)
+        self.time_eq(new_cutoff, now)
+
+    def test_catch_up_from(self):
+        # catch_up_from() asks Overdrive about recent changes by
+        # calling recently_changed_ids(), and mirrors those changes
+        # locally by calling update_licensepool().
+        class MockAPI(object):
+            def __init__(self, *ignore, **kwignore):
+                self.licensepools = []
+                self.update_licensepool_calls = []
+
+            def update_licensepool(self, book_id):
+                pool, is_new, is_changed = self.licensepools.pop(0)
+                self.update_licensepool_calls.append((book_id, pool))
+                return pool, is_new, is_changed
+
+        class MockMonitor(OverdriveCirculationMonitor):
+
+            recently_changed_ids_called_with = None 
+            def recently_changed_ids(self, start, cutoff):
+                self.recently_changed_ids_called_with = (start, cutoff)
+                return [1, 2, 3, 4]
+
+        monitor = MockMonitor(self._db, self.collection, api_class=MockAPI)
+        monitor.maximum_consecutive_unchanged_books = 2
+        api = monitor.api
+
+        # The 'Overdrive API' is ready to tell us about four books,
+        # but only one of them (the first) represents a change from what
+        # we already know.
+        lp1 = self._licensepool(None)
+        lp2 = self._licensepool(None)
+        lp3 = self._licensepool(None)
+        lp4 = object()
+        api.licensepools.append((lp1, True, True))
+        api.licensepools.append((lp2, False, False))
+        api.licensepools.append((lp3, False, False))
+        api.licensepools.append(lp4)
+
+        monitor.catch_up_from(object(), object(), object())
+
+        # We called update_licensepool on the first three books,
+        # and got the first three LicensePools from the queue.
+        eq_([(1, lp1),(2, lp2),(3, lp3)], api.update_licensepool_calls)
+
+        # At that point we gave up because we'd gotten two consecutive
+        # books from Overdrive that contained no new information.
```

Comment:
When you say "we gave up", does that mean that it just stops monitor at that point? What happens to the fourth LicensePool that is still in `api.licensepools`?

Reply:
Nothing happens to that LicensePool. It represents a missed opportunity.

When we run this monitor for real, it will periodically give us 20 or 30 consecutive "recent" books that, upon inspection, haven't actually changed. But then we'll get a chunk of books that _have_ actually changed. So we can't stop looking as soon as we find a book that didn't actually change. The data we get from Overdrive isn't well-ordered, the way it is when we get it from Bibliotheca.

At the same time, we can't keep going on forever just in case there's a book way down in the "recent" stack that has new data. (That does happen, but we use a different monitor to catch those changes.) As a compromise, we stop looking at the list of "recent" books once we see 100 consecutive "recent" books that haven't changed.